src/Cello/test_CelloArray.cpp

```diff
@@ -0,0 +1,1044 @@
+// See LICENSE_CELLO file for license and copyright information
+
+/// @file     test_CelloArray.cpp
+/// @author   Matthew Abruzzo (matthewabruzzo@gmail.com)
+/// @date     2020-06-09
+/// @brief    Test program for the CelloArray class template
+///
+/// The main testing strategy here is to test some of the core functionallity
+/// of CelloArray with:
+///    - SimpleInitializationTests
+///    - SimpleAssignmentTests
+///    - SizeShapeTests
+/// And then use more sophisticated machinery that relies on this
+/// functionallity to test the remaining features.
+
+#include <cstdlib>
+#include <initializer_list>
+#include <vector>
+
+#include "main.hpp"
+#include "test.hpp"
+#include "array.hpp"
+
+//----------------------------------------------------------------------
+
+template<typename T, std::size_t D>
+class MemManagedArrayBuilder{
+  // This template class is used to build and a manage the lifetime of a
+  // CelloArray that manages its own memory
+  //
+  // In the future it might be a good idea to have this and
+  // PtrWrapArrayBuilder have a shared base class
+public:
+  template<typename... Args>
+  MemManagedArrayBuilder(Args... args) : arr_(args...) {}
+  CelloArray<T,D>* get_arr() { return &arr_; }
+  T* get_wrapped_ptr() { return NULL; }
+  static std::string name() { return "MemManagedArrayBuilder"; }
+
+private:
+  CelloArray<T,D> arr_;
+};
+
+//----------------------------------------------------------------------
+
+template<typename T, std::size_t D>
+class PtrWrapArrayBuilder{
+  // This template class is used to build and a manage the lifetime of a
+  // CelloArray wraps an existing pointer. This template class also manages the
+  // lifetime of the underlying pointer.
+  //
+  // In the future it might be a good idea to have this and
+  // MemManagedArrayBuilder have a shared base class
+public:
+
+  template<typename... Args>
+  PtrWrapArrayBuilder(Args... args)
+    : arr_ptr_(nullptr),
+      ptr_(nullptr)
+  {
+    // we are sort of cheating here to compute the required size
+    CelloArray<T,D> temp(args...);
+    ptr_ = new T [temp.size()]{};
+    arr_ptr_ = new CelloArray<T,D>(ptr_,args...);
+  }
+
+  ~PtrWrapArrayBuilder() {
+    delete arr_ptr_;
+    delete[] ptr_;
+  }
+
+  CelloArray<T,D>* get_arr() {return arr_ptr_;}
```

Comment:
I looked into this, and I tend to think it is overkill here primarily because everytime we call this function, we would need to lock the pointer immediately to make it into a shared_ptr.

I might make this into a shared_ptr instead

Reply:
Ok, so I thought about this some more and this is almost certainly not worth doing

I'm stating the obvious here, but the primary reason why we would want to use a shared_ptr is so that we can avoid having a dangling pointer to a `CelloArray` if the `PtrWrapArrayBuilder` object is deallocated first
- The problem with this is that the `CelloArray` is wrapping a pointer whose lifetime is the same as `PtrWrapArrayBuilder`. Thus, the `CelloArray` pointer returned by `get_arr()` can only be valid for as long as `PtrWrapArrayBuilder` exists
- In theory, we could take advantage of `std::shared_ptr`'s ability to manage the lifetime of a separate pointer than the one it stores, but this would involve introducing another layer of abstraction and I definitely think that would be overkill. For the sake of example, I've attached the block of code that I think would be required to do this properly

Considering this is just a unit test, I favor the much shorter, easier to understand implementation that we already have

```C++
template<typename T, std::size_t D>
struct PtrArrayPair{
  template<typename... Args>
  PtrArrayPair(Args... args){
    // we are sort of cheating here to compute the required size
    CelloArray<T,D> temp(args...);
    ptr = new T [temp.size()]{};
    arr_ptr = new CelloArray<T,D>(ptr_,args...);
  }

  ~PtrArrayPair(){
    delete arr_ptr;
    delete[] ptr;
  }

  // public attributes:
  CelloArray<T,D>* arr_ptr;
  T* ptr;
};

template<typename T, std::size_t D>
class PtrWrapArrayBuilder{
  // This template class is used to build and a manage the lifetime of a
  // CelloArray wraps an existing pointer. This template class also manages the
  // lifetime of the underlying pointer.
  //
  // In the future it might be a good idea to have this and
  // MemManagedArrayBuilder have a shared base class
public:

  template<typename... Args>
  PtrWrapArrayBuilder(Args... args)
  {
    data_pair_ = std::make_shared<PtrArrayPair<T,D>>(args...);
  }

  std::shared_ptr<CelloArray<T,D>> get_arr()
  {
    return std::shared_ptr<CelloArray<T,D>>(data_pair_,
                                            data_pair_->arr_ptr);
  }

  std::shared_ptr<T> get_wrapped_ptr() {
    return std::shared_ptr<T>(data_pair_, data_pair_->ptr);
  }

  static std::string name() { return "PtrWrapArrayBuilder"; }

private:
  std::shared_ptr<PtrArrayPair<T,D>> data_pair_;
};
```